sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -148,6 +148,40 @@ object Cast {
     case _ => false
   }
 
+  /**
+   * Returns true iff we can safely up-cast the `from` type to `to` type without any truncating or
+   * precision lose or possible runtime failures. For example, long -> int, string -> int are not
+   * up-cast.
+   */
+  def canUpCast(from: DataType, to: DataType): Boolean = (from, to) match {
```

Comment:
Where is this from? @bersprockets Can you note the original PR of this in the PR description?

Reply:
I think this should have some tests in the original PR. Should we backport them together? @cloud-fan @dongjoon-hyun

Reply:
Usually, we don't backport the original PR if this kind of requirement arises. How do you think about this backport, @cloud-fan and @gatorsmile ?

This PR needs `canUpCast` (here) and `sameOutput` (https://github.com/apache/spark/pull/25068/files#diff-27c76f96a7b2733ecfd6f46a1716e153R140) additionally.

Reply:
The rule `AliasViewChild` is already there before we have `canUpCast` and `sameOutput`. We just need to update the corresponding code to the old version without `canUpCast` and `sameOutput`.

Reply:
+1 for @cloud-fan 's comment. Thanks!